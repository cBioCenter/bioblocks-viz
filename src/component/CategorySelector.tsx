import * as React from 'react';
import { Dropdown } from 'semantic-ui-react';

export interface ICategorySelectorProps {
  /** Categories to select from. */
  categories: string[];
  /** Callback for when a new category is selected. */
  onCategoryChange?: (event: React.SyntheticEvent<any>, data: object) => void;
}

/**
 * Class to represent a dropdown.
 *
 * @extends {React.Component<ICategorySelectorProps, any>}
 */
export default class CategorySelector extends React.Component<ICategorySelectorProps, any> {
  constructor(props: ICategorySelectorProps) {
    super(props);
  }

  public render() {
    return (
      this.props.categories && (
        <Dropdown
          fluid={true}
          onChange={this.props.onCategoryChange}
          options={[
            { key: 'all', text: 'all', value: undefined },
            ...this.props.categories.map(cat => {
              return { key: cat, text: cat, value: cat };
            }),
          ]}
          placeholder={'Select Category'}
          search={true}
        />
      )
    );
  }
}

export { CategorySelector };
