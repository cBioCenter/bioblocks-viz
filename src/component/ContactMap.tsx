import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { ButtonProps, Icon } from 'semantic-ui-react';
import { createContainerActions, createResiduePairActions } from '~bioblocks-viz~/action';
import {
  ComponentCard,
  ContactMapChart,
  generateChartDataEntry,
  IComponentMenuBarItem,
  IContactMapChartData,
  IContactMapChartPoint,
} from '~bioblocks-viz~/component';
import {
  BIOBLOCKS_CSS_STYLE,
  BioblocksChartEvent,
  BioblocksWidgetConfig,
  ButtonGroupWidgetConfig,
  CONFIGURATION_COMPONENT_TYPE,
  CouplingContainer,
  IContactMapData,
  ICouplingScore,
  RESIDUE_TYPE,
  SECONDARY_STRUCTURE,
  SECONDARY_STRUCTURE_KEYS,
  SECONDARY_STRUCTURE_SECTION,
} from '~bioblocks-viz~/data';
import { ColorMapper, EMPTY_FUNCTION, generateCouplingScoreHoverText } from '~bioblocks-viz~/helper';
import { ILockedResiduePair } from '~bioblocks-viz~/reducer';
import { getCandidates, getHovered, getLocked, selectCurrentItems } from '~bioblocks-viz~/selector';

export type CONTACT_MAP_CB_RESULT_TYPE = ICouplingScore;
export type ContactMapCallback = (coupling: CONTACT_MAP_CB_RESULT_TYPE) => void;

export interface IContactMapProps {
  candidateResidues: RESIDUE_TYPE[];
  configurations: BioblocksWidgetConfig[];
  data: IContactMapData;
  formattedPoints: IContactMapChartData[];
  height: number | string;
  highlightColor: string;
  hoveredResidues: RESIDUE_TYPE[];
  hoveredSecondaryStructures: SECONDARY_STRUCTURE_SECTION[];
  isDataLoading: boolean;
  lockedResiduePairs: ILockedResiduePair;
  observedColor: string;
  secondaryStructureColors?: ColorMapper<SECONDARY_STRUCTURE_KEYS>;
  selectedSecondaryStructures: SECONDARY_STRUCTURE;
  showConfigurations: boolean;
  style?: BIOBLOCKS_CSS_STYLE;
  width: number | string;
  addHoveredResidues(residues: RESIDUE_TYPE[]): void;
  addHoveredSecondaryStructure(section: SECONDARY_STRUCTURE_SECTION): void;
  addSelectedSecondaryStructure(section: SECONDARY_STRUCTURE_SECTION): void;
  removeAllLockedResiduePairs(): void;
  removeAllSelectedSecondaryStructures(): void;
  removeSecondaryStructure(section: SECONDARY_STRUCTURE_SECTION): void;
  removeHoveredResidues(): void;
  removeHoveredSecondaryStructure(section: SECONDARY_STRUCTURE_SECTION): void;
  onBoxSelection?(residues: RESIDUE_TYPE[]): void;
  toggleLockedResiduePair(residuePair: ILockedResiduePair): void;
}

export const initialContactMapState = {
  pointsToPlot: new Array<IContactMapChartData>(),
};

export type ContactMapState = Readonly<typeof initialContactMapState>;

export class ContactMapClass extends React.Component<IContactMapProps, ContactMapState> {
  public static defaultProps = {
    addHoveredResidues: EMPTY_FUNCTION,
    addHoveredSecondaryStructure: EMPTY_FUNCTION,
    addSelectedSecondaryStructure: EMPTY_FUNCTION,
    candidateResidues: [],
    configurations: new Array<BioblocksWidgetConfig>(),
    data: {
      couplingScores: new CouplingContainer(),
      pdbData: { experimental: undefined, predicted: undefined },
      secondaryStructures: new Array<SECONDARY_STRUCTURE>(),
    },
    enableSliders: true,
    formattedPoints: new Array<IContactMapChartData>(),
    height: '100%',
    highlightColor: '#ff8800',
    hoveredResidues: [],
    hoveredSecondaryStructures: [],
    isDataLoading: false,
    lockedResiduePairs: {},
    observedColor: '#0000ff',
    removeAllLockedResiduePairs: EMPTY_FUNCTION,
    removeAllSelectedSecondaryStructures: EMPTY_FUNCTION,
    removeHoveredResidues: EMPTY_FUNCTION,
    removeHoveredSecondaryStructure: EMPTY_FUNCTION,
    removeSecondaryStructure: EMPTY_FUNCTION,
    selectedSecondaryStructures: [],
    showConfigurations: true,
    toggleLockedResiduePair: EMPTY_FUNCTION,
    width: '100%',
  };

  public readonly state: ContactMapState = initialContactMapState;

  constructor(props: IContactMapProps) {
    super(props);
  }

  public componentDidMount() {
    this.setupPointsToPlot(this.props.data.couplingScores);
  }

  public componentDidUpdate(prevProps: IContactMapProps) {
    const { data, lockedResiduePairs } = this.props;
    if (data !== prevProps.data || lockedResiduePairs !== prevProps.lockedResiduePairs) {
      this.setupPointsToPlot(data.couplingScores);
    }
  }

  public onNodeSizeChange = (index: number, nodeSizeMod: number) => (
    event?: React.MouseEvent<HTMLButtonElement>,
    data?: ButtonProps,
  ) => {
    const { pointsToPlot } = this.state;

    this.setState({
      pointsToPlot: [
        ...pointsToPlot.slice(0, index),
        {
          ...pointsToPlot[index],
          // This ensures a number in the range [1, 10]. 0 is an invalid point size in plotly.
          nodeSize: Math.min(Math.max(Math.abs(pointsToPlot[index].nodeSize + nodeSizeMod), 1), 10),
        },
        ...pointsToPlot.slice(index + 1),
      ],
    });
  };

  public render() {
    const { pointsToPlot } = this.state;

    return this.renderContactMapChart(pointsToPlot);
  }

  /**
   * Given a chart data entry, sets the node size if one already is set for this point.
   *
   * @param chartDatum A single chart data entry.
   */
  protected applyNodeSize = (chartDatum: IContactMapChartData) => {
    const { pointsToPlot } = this.state;
    const index = pointsToPlot.findIndex(currentPoint => currentPoint.name === chartDatum.name);
    if (index >= 0) {
      return {
        ...chartDatum,
        nodeSize: pointsToPlot[index].nodeSize,
      };
    } else {
      return chartDatum;
    }
  };

  /**
   * Returns an {i, j} pair for a hovered residue. Handles case when a single residue is hovered.
   *
   * @param hoveredResidues Array of residues that are hovered.
   */
  protected generateHoveredResiduePairs = (hoveredResidues: RESIDUE_TYPE[]) => {
    if (hoveredResidues.length >= 1) {
      return [
        {
          i: hoveredResidues[0],
          j: hoveredResidues.length === 1 ? hoveredResidues[0] : hoveredResidues[1],
        },
      ];
    }

    return [];
  };

  /**
   * Returns contact map points for locked residue pairs.
   *
   * @param hoveredResidues Array of residues that are hovered.
   */
  protected generateLockedResiduePairs = (lockedResiduePairs: ILockedResiduePair) => {
    const lockedResiduePairKeys = Object.keys(lockedResiduePairs);

    return lockedResiduePairKeys.reduce((reduceResult: IContactMapChartPoint[], key) => {
      const keyPair = lockedResiduePairs[key];
      if (keyPair && keyPair.length === 2) {
        reduceResult.push({ i: keyPair[0], j: keyPair[1], dist: 0 });
      }

      return reduceResult;
    }, new Array<IContactMapChartPoint>());
  };

  protected generateSelectedResiduesChartData = (
    highlightColor: string,
    chartName: string,
    nodeSize: number,
    chartPoints: IContactMapChartPoint[],
  ) => {
    return generateChartDataEntry('none', highlightColor, chartName, '', nodeSize, chartPoints, {
      marker: {
        color: new Array<string>(chartPoints.length * 2).fill(highlightColor),
        line: {
          color: highlightColor,
          width: 3,
        },
        symbol: 'circle-open',
      },
    });
  };

  /**
   * Gets the color from the provided chart entry.
   *
   * The following are checked in order - the first one found is the color returned:
   * - Marker's line color.
   * - Marker's color.
   * - Marker's colorscale.
   * - Line color.
   */
  protected getColorFromEntry = (entry: IContactMapChartData) => {
    if (entry.marker) {
      if (entry.marker.line && entry.marker.line.color) {
        return entry.marker.line.color;
      }
      if (entry.marker.color) {
        return Array.isArray(entry.marker.color) ? entry.marker.color[0] : entry.marker.color;
      }
      if (entry.marker.colorscale) {
        return Array.isArray(entry.marker.colorscale) ? entry.marker.colorscale[0][1] : entry.marker.colorscale;
      }
    }
    if (entry.line && entry.line.color) {
      return entry.line.color;
    }
  };

  protected getDockConfigs = () => [
    {
      isVisibleCb: () =>
        this.props.selectedSecondaryStructures.length >= 1 || Object.values(this.props.lockedResiduePairs).length >= 1,
      onClick: () => {
        this.props.removeAllLockedResiduePairs();
        this.props.removeAllSelectedSecondaryStructures();
        this.props.removeHoveredResidues();
      },
      text: 'Clear Selections',
    },
  ];

  protected getMenuConfigs = (
    Filters: BioblocksWidgetConfig[],
    pointsToPlot: IContactMapChartData[],
  ): IComponentMenuBarItem[] => [
    {
      component: {
        configs: { Filters },
        name: 'POPUP',
      },
      description: 'Filter',
      iconName: 'filter',
    },
    {
      component: {
        configs: this.getNodeSizeSliderConfigs(pointsToPlot),
        name: 'POPUP',
      },
      description: 'Settings',
    },
  ];

  protected getNodeSizeSliderConfigs = (entries: IContactMapChartData[]) => ({
    'Node Sizes': entries.map(
      (entry, index): ButtonGroupWidgetConfig => {
        const color = this.getColorFromEntry(entry);

        // prettier-ignore
        const options = [(
          <Icon
            key={`node-size-slider-${index}-minus`}
            name={'minus'}
            onClick={this.onNodeSizeChange(index, -1)}
            size={'small'}
            style={{ color }}
          />), (
          <Icon
            key={`node-size-slider-${index}-2`}
            name={'circle'}
            size={'small'}
            style={{ color }}
          />), (
          <Icon
            key={`node-size-slider-${index}-plus`}
            name={'plus'}
            onClick={this.onNodeSizeChange(index, 1)}
            size={'small'}
            style={{ color }}
          />
        )];

        return {
          id: `node-size-slider-${index}`,
          name: entry.name,
          options,
          style: { padding: '.5em' },
          type: CONFIGURATION_COMPONENT_TYPE.BUTTON_GROUP,
        };
      },
    ),
  });

  protected handleAxisClick = (e: BioblocksChartEvent) => {
    const { addSelectedSecondaryStructure, data, removeSecondaryStructure, selectedSecondaryStructures } = this.props;

    for (const secondaryStructure of data.secondaryStructures) {
      for (const section of secondaryStructure) {
        if (section.contains(...e.selectedPoints)) {
          if (selectedSecondaryStructures.includes(section)) {
            removeSecondaryStructure(section);
          } else {
            addSelectedSecondaryStructure(section);
          }
        }
      }
    }
  };

  protected handleAxisEnter = (e: BioblocksChartEvent) => {
    const { addHoveredSecondaryStructure, hoveredSecondaryStructures, data } = this.props;

    for (const secondaryStructure of data.secondaryStructures) {
      for (const section of secondaryStructure) {
        if (
          section.contains(...e.selectedPoints) &&
          !hoveredSecondaryStructures.find(secStruct => section.isEqual(secStruct))
        ) {
          addHoveredSecondaryStructure(section);
        }
      }
    }
  };

  protected handleAxisLeave = (e: BioblocksChartEvent) => {
    const { data, hoveredSecondaryStructures, removeHoveredSecondaryStructure } = this.props;

    for (const secondaryStructure of data.secondaryStructures) {
      for (const section of secondaryStructure) {
        const searchResult = hoveredSecondaryStructures.find(secStruct => section.isEqual(secStruct));
        if (section.contains(...e.selectedPoints) && searchResult) {
          removeHoveredSecondaryStructure(searchResult);
        }
      }
    }
  };

  protected onMouseClick = (cb: (residues: ILockedResiduePair) => void) => (e: BioblocksChartEvent) => {
    if (e.isAxis()) {
      this.handleAxisClick(e);
    } else {
      cb({ [e.selectedPoints.sort().toString()]: e.selectedPoints });
    }
  };

  protected onMouseEnter = (cb: (residue: RESIDUE_TYPE[]) => void) => (e: BioblocksChartEvent) => {
    if (e.isAxis()) {
      this.handleAxisEnter(e);
    } else {
      cb(e.selectedPoints);
    }
  };

  protected onMouseLeave = (cb?: (residue: RESIDUE_TYPE[]) => void) => (e: BioblocksChartEvent) => {
    if (e.isAxis()) {
      this.handleAxisLeave(e);
    } else if (cb) {
      cb(e.selectedPoints);
    }
  };

  protected onMouseSelect = (cb?: (residues: RESIDUE_TYPE[]) => void) => (e: BioblocksChartEvent) => {
    if (cb) {
      // For the contact map, all the x/y values are mirrored and correspond directly with i/j values.
      // Thus, all the residue numbers can be obtained by getting either all x or values from ths selected points.
      cb(e.selectedPoints.map(point => point));
    }
  };

  protected renderContactMapChart(pointsToPlot: IContactMapChartData[]) {
    const {
      addHoveredResidues,
      candidateResidues,
      configurations,
      data,
      height,
      isDataLoading,
      onBoxSelection,
      removeHoveredResidues,
      secondaryStructureColors,
      selectedSecondaryStructures,
      showConfigurations,
      toggleLockedResiduePair,
      width,
    } = this.props;

    const range = data.couplingScores.residueIndexRange.max + data.couplingScores.residueIndexRange.min;

    return (
      <ComponentCard
        componentName={'Contact Map'}
        dockItems={this.getDockConfigs()}
        isDataReady={data.couplingScores.allContacts.length >= 1}
        menuItems={this.getMenuConfigs(configurations, pointsToPlot)}
      >
        <ContactMapChart
          candidateResidues={candidateResidues}
          contactData={pointsToPlot}
          height={height}
          isDataLoading={isDataLoading}
          onClickCallback={this.onMouseClick(toggleLockedResiduePair)}
          onHoverCallback={this.onMouseEnter(addHoveredResidues)}
          onSelectedCallback={this.onMouseSelect(onBoxSelection)}
          onUnHoverCallback={this.onMouseLeave(removeHoveredResidues)}
          range={range}
          secondaryStructures={data.secondaryStructures ? data.secondaryStructures : []}
          secondaryStructureColors={secondaryStructureColors}
          selectedSecondaryStructures={[selectedSecondaryStructures]}
          showConfigurations={showConfigurations}
          width={width}
        />
      </ComponentCard>
    );
  }

  protected setupPointsToPlot(couplingContainer: CouplingContainer) {
    const { data, lockedResiduePairs, hoveredResidues, formattedPoints, observedColor, highlightColor } = this.props;

    const chartNames = {
      selected: 'Selected Residue Pair',
      structure: `${data.pdbData ? (data.pdbData.experimental ? 'X-ray' : 'Inferred') : 'Unknown'} Structure Contact`,
    };

    const observedContactPoints = couplingContainer.getObservedContacts();
    const result = new Array<IContactMapChartData>(
      generateChartDataEntry(
        'text',
        { start: observedColor, end: 'rgb(100,177,200)' },
        `Structure Contact (${data.pdbData ? 'PDB' : 'Coupling Scores'})`,
        chartNames.structure,
        4,
        observedContactPoints,
        {
          text: observedContactPoints.map(generateCouplingScoreHoverText),
        },
      ),
      ...formattedPoints,
    );

    const chartPoints = new Array<IContactMapChartPoint>();

    chartPoints.push(...this.generateHoveredResiduePairs(hoveredResidues));
    chartPoints.push(...this.generateLockedResiduePairs(lockedResiduePairs));

    result.push(this.generateSelectedResiduesChartData(highlightColor, chartNames.selected, 4, chartPoints));

    this.setState({
      ...this.state,
      pointsToPlot: result.map(this.applyNodeSize),
    });
  }
}

const mapStateToProps = (state: { [key: string]: any }) => ({
  candidateResidues: getCandidates(state).toArray(),
  hoveredResidues: getHovered(state).toArray(),
  hoveredSecondaryStructures: selectCurrentItems<SECONDARY_STRUCTURE_SECTION>(
    state,
    'secondaryStructure/hovered',
  ).toArray(),
  lockedResiduePairs: getLocked(state).toJS(),
  selectedSecondaryStructures: selectCurrentItems<SECONDARY_STRUCTURE_SECTION>(
    state,
    'secondaryStructure/selected',
  ).toArray(),
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      addHoveredResidues: createResiduePairActions().hovered.set,
      addHoveredSecondaryStructure: createContainerActions('secondaryStructure/hovered').add,
      addSelectedSecondaryStructure: createContainerActions('secondaryStructure/selected').add,
      removeAllLockedResiduePairs: createResiduePairActions().locked.clear,
      removeAllSelectedSecondaryStructures: createContainerActions('secondaryStructure/selected').clear,
      removeHoveredResidues: createResiduePairActions().hovered.clear,
      removeHoveredSecondaryStructure: createContainerActions('secondaryStructure/hovered').remove,
      removeSecondaryStructure: createContainerActions('secondaryStructure/selected').remove,
      toggleLockedResiduePair: createResiduePairActions().locked.toggle,
    },
    dispatch,
  );

export const ContactMap = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ContactMapClass);
