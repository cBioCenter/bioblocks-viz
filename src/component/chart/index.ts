export * from './AuxiliaryAxis';
export * from './ContactMapChart';
export * from './PlotlyChart';
export * from './SecondaryStructureAxis';
