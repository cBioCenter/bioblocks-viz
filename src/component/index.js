"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./chart"), exports);
tslib_1.__exportStar(require("./widget"), exports);
tslib_1.__exportStar(require("./ContactMapComponent"), exports);
tslib_1.__exportStar(require("./FeatureViewer"), exports);
tslib_1.__exportStar(require("./NGLComponent"), exports);
tslib_1.__exportStar(require("./TensorTComponent"), exports);
tslib_1.__exportStar(require("./UMAPVisualization"), exports);
//# sourceMappingURL=index.js.map