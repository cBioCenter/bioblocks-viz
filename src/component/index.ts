export * from './chart';
export * from './widget';

export * from './ContactMap';
export * from './FeatureViewer';
export * from './InfoPanel';
export * from './MorpheusComponent';
export * from './NGLComponent';
export * from './PredictedContactMap';
export * from './TComponent';
export * from './TensorTComponent';
export * from './TFrameComponent';
export * from './VizSelectorPanel';
