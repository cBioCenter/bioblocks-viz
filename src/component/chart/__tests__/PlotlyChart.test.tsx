import { mount } from 'enzyme';
import * as React from 'react';

import { IPlotlyChartProps, PlotlyChart } from '~chell-viz~/component';
import { CHELL_CHART_EVENT_TYPE, CHELL_CHART_PIECE, ChellChartEvent, IPlotlyData } from '~chell-viz~/data';
import {
  dispatchPlotlyEvent,
  dispatchPlotlySecondaryAxisEvent,
  dispatchPlotlySelectionEvent,
  IMockDict,
} from '~chell-viz~/test';

beforeEach(() => {
  jest.resetModules();
});

describe('PlotlyChart', () => {
  let sampleData: Array<Partial<IPlotlyData>>;
  beforeEach(() => {
    sampleData = [
      {
        marker: {
          color: 'blue',
        },
        mode: 'markers',
        type: 'pointcloud',
        xy: new Float32Array([1, 2, 3, 4]),
      },
    ];
  });

  /**
   * Helper function to create and wait for a PlotlyChart to be mounted.
   *
   * @param props Custom props to be passed to the chart.
   * @returns A wrapper for the PlotlyChart that has been mounted.
   */
  const getMountedPlotlyChart = async (props: IPlotlyChartProps) => {
    const wrapper = mount(<PlotlyChart {...props} />);
    wrapper.update();
    await Promise.resolve();

    return wrapper;
  };

  it('Should match existing snapshot when given empty data.', () => {
    const wrapper = mount(<PlotlyChart data={[]} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should match existing snapshot when given sample data.', () => {
    const wrapper = mount(<PlotlyChart data={sampleData} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Event Callbacks', () => {
    it('Should handle plotly events.', async () => {
      const spies: IMockDict = {
        onClickSpy: jest.fn(),
        onDoubleClickSpy: jest.fn(),
        onHoverSpy: jest.fn(),
        onRelayoutSpy: jest.fn(),
        onSelectedSpy: jest.fn(),
        onUnHoverSpy: jest.fn(),
      };

      const wrapper = await getMountedPlotlyChart({
        data: sampleData,
        onClickCallback: spies.onClickSpy,
        onDoubleClickCallback: spies.onDoubleClickSpy,
        onHoverCallback: spies.onHoverSpy,
        onRelayoutCallback: spies.onRelayoutSpy,
        onSelectedCallback: spies.onSelectedSpy,
        onUnHoverCallback: spies.onUnHoverSpy,
      });

      await Promise.all([
        dispatchPlotlyEvent(wrapper, 'plotly_click'),
        dispatchPlotlyEvent(wrapper, 'plotly_doubleclick'),
        dispatchPlotlyEvent(wrapper, 'plotly_hover'),
        dispatchPlotlyEvent(wrapper, 'plotly_relayout'),
        dispatchPlotlySelectionEvent(wrapper),
        dispatchPlotlyEvent(wrapper, 'plotly_unhover'),
      ]);

      for (const key of Object.keys(spies)) {
        expect(spies[key]).toHaveBeenCalledTimes(1);
      }
    });

    it('Should attach events when the container is attached.', async () => {
      const spies: IMockDict = {
        onClickSpy: jest.fn(),
        onSelectedSpy: jest.fn(),
      };

      const wrapper = await getMountedPlotlyChart({
        data: sampleData,
        onClickCallback: spies.onClickSpy,
        onSelectedCallback: spies.onSelectedSpy,
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('Should call the appropriate callback when the window is resized.', async () => {
      const onResizeSpy = jest.fn();

      const wrapper = await getMountedPlotlyChart({ data: sampleData });
      const chartInstance = wrapper.instance() as PlotlyChart;

      chartInstance.resize = onResizeSpy;
      chartInstance.attachListeners();
      window.dispatchEvent(new Event('resize'));
      expect(onResizeSpy).toHaveBeenCalledTimes(1);
    });

    it('Should call the appropriate callback when plotly emits a click event.', async () => {
      const onClickSpy = jest.fn();
      const wrapper = await getMountedPlotlyChart({
        data: sampleData,
        onClickCallback: onClickSpy,
      });
      await dispatchPlotlyEvent(wrapper, 'plotly_click');
      expect(onClickSpy).toBeCalled();
    });

    it('Should return the appropriate event when plotly emits a click event on a point.', async () => {
      const onClickSpy = jest.fn();
      const wrapper = await getMountedPlotlyChart({
        data: sampleData,
        onClickCallback: onClickSpy,
      });
      await dispatchPlotlyEvent(wrapper, 'plotly_click', { x: 1, y: 2 });
      const chellEvent = onClickSpy.mock.calls[0][0] as ChellChartEvent;
      expect(chellEvent.chartPiece).toBe(CHELL_CHART_PIECE.POINT);
      expect(chellEvent.type).toBe(CHELL_CHART_EVENT_TYPE.CLICK);
      expect(chellEvent.selectedPoints).toEqual([1, 2]);
    });

    it('Should return the appropriate event when plotly emits a click event on a secondary x axis.', async () => {
      const onClickSpy = jest.fn();
      const wrapper = await getMountedPlotlyChart({
        data: sampleData,
        onClickCallback: onClickSpy,
      });
      dispatchPlotlySecondaryAxisEvent(wrapper, 'plotly_click', { data: { xaxis: 'x2', yaxis: 'y' }, x: 1, y: 2 });
      const chellEvent = onClickSpy.mock.calls[0][0] as ChellChartEvent;
      expect(chellEvent.chartPiece).toBe(CHELL_CHART_PIECE.AXIS);
      expect(chellEvent.type).toBe(CHELL_CHART_EVENT_TYPE.CLICK);
      expect(chellEvent.selectedPoints).toEqual([2]);
    });

    it('Should return the appropriate event when plotly emits a click event on a secondary y axis.', async () => {
      const onClickSpy = jest.fn();
      const wrapper = await getMountedPlotlyChart({
        data: sampleData,
        onClickCallback: onClickSpy,
      });
      dispatchPlotlySecondaryAxisEvent(wrapper, 'plotly_click', { data: { xaxis: 'x', yaxis: 'y2' }, x: 1, y: 2 });
      const chellEvent = onClickSpy.mock.calls[0][0] as ChellChartEvent;
      expect(chellEvent.chartPiece).toBe(CHELL_CHART_PIECE.AXIS);
      expect(chellEvent.type).toBe(CHELL_CHART_EVENT_TYPE.CLICK);
      expect(chellEvent.selectedPoints).toEqual([1]);
    });

    it('Should call the appropriate callback when plotly emits a hover event.', async () => {
      const onHoverSpy = jest.fn();
      const wrapper = await getMountedPlotlyChart({
        data: sampleData,
        onHoverCallback: onHoverSpy,
      });
      await dispatchPlotlyEvent(wrapper, 'plotly_hover');
      expect(onHoverSpy).toBeCalled();
    });

    it('Should call the appropriate callback when plotly emits a selected event.', async () => {
      const onSelectedSpy = jest.fn();
      const wrapper = await getMountedPlotlyChart({
        data: sampleData,
        onSelectedCallback: onSelectedSpy,
      });
      await dispatchPlotlySelectionEvent(wrapper);
      expect(onSelectedSpy).toBeCalled();
    });

    it('Should call the appropriate callback when plotly emits a selected event for a range of points.', async () => {
      const onSelectedSpy = jest.fn();
      const wrapper = await getMountedPlotlyChart({
        data: sampleData,
        onSelectedCallback: onSelectedSpy,
      });
      await dispatchPlotlySelectionEvent(wrapper, { points: [], range: { x: [0, 1], y: [0, 1] } });
      expect(onSelectedSpy).toBeCalled();
    });
  });

  it('Should unmount correctly.', async () => {
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
    });

    const chartInstance = wrapper.instance() as PlotlyChart;
    expect(chartInstance.plotlyCanvas).not.toBeNull();
    expect(chartInstance.plotlyCanvas).not.toBeUndefined();
    wrapper.unmount();
    expect(chartInstance.plotlyCanvas).toBeNull();
  });

  it('Should not call draw if data is unchanged.', async () => {
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
    });

    const chartInstance = wrapper.instance() as PlotlyChart;
    const drawSpy = jest.fn();
    chartInstance.draw = drawSpy;

    wrapper.setProps({
      data: sampleData,
    });

    expect(drawSpy).toHaveBeenCalledTimes(0);
  });

  it('Should call draw when data is updated.', async () => {
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
    });

    const chartInstance = wrapper.instance() as PlotlyChart;
    const drawSpy = jest.fn();
    chartInstance.draw = drawSpy;

    wrapper.setProps({
      data: {},
    });

    expect(drawSpy).toHaveBeenCalledTimes(1);
  });

  it('Should call draw when layout is updated.', async () => {
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
    });

    const chartInstance = wrapper.instance() as PlotlyChart;
    const drawSpy = jest.fn();
    chartInstance.draw = drawSpy;

    wrapper.setProps({
      layout: {
        autosize: false,
      },
    });

    expect(drawSpy).toHaveBeenCalledTimes(1);
  });

  it('Should call draw when config is updated.', async () => {
    const wrapper = await getMountedPlotlyChart({
      data: sampleData,
    });

    const chartInstance = wrapper.instance() as PlotlyChart;
    const drawSpy = jest.fn();
    chartInstance.draw = drawSpy;

    wrapper.setProps({
      config: {
        displayModeBar: true,
      },
    });

    expect(drawSpy).toHaveBeenCalledTimes(1);
  });

  it('Should handle extra axes correctly.', async () => {
    const wrapper = await getMountedPlotlyChart({
      data: [...sampleData, ...[{ xaxis: 'x2' }, { yaxis: 'y2' }]],
    });

    expect(wrapper).toMatchSnapshot();
  });
});
