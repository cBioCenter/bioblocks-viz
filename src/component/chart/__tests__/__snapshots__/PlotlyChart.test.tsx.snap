// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlotlyChart Should attach events when the container is attached. 1`] = `
<PlotlyChart
  config={Object {}}
  data={
    Array [
      Object {
        "marker": Object {
          "color": "blue",
        },
        "mode": "markers",
        "type": "pointcloud",
        "xy": Float32Array [
          1,
          2,
          3,
          4,
        ],
      },
    ]
  }
  height="100%"
  layout={Object {}}
  onClickCallback={[MockFunction]}
  onSelectedCallback={[MockFunction]}
  showLoader={true}
  width="100%"
>
  <div>
    <Dimmer
      active={false}
    >
      <DimmerInner
        active={false}
      >
        <div
          className="ui dimmer"
          onClick={[Function]}
        >
          <div
            className="content"
          >
            <Loader>
              <div
                className="ui loader"
              />
            </Loader>
          </div>
        </div>
      </DimmerInner>
    </Dimmer>
    <div
      className="plotly-chart"
      style={
        Object {
          "height": "100%",
          "marginBottom": 5,
          "width": "100%",
        }
      }
    />
  </div>
</PlotlyChart>
`;

exports[`PlotlyChart Should handle extra axes correctly. 1`] = `
<PlotlyChart
  config={Object {}}
  data={
    Array [
      Object {
        "marker": Object {
          "color": "blue",
        },
        "mode": "markers",
        "type": "pointcloud",
        "xy": Float32Array [
          1,
          2,
          3,
          4,
        ],
      },
      Object {
        "xaxis": "x2",
      },
      Object {
        "yaxis": "y2",
      },
    ]
  }
  height="100%"
  layout={Object {}}
  showLoader={true}
  width="100%"
>
  <div>
    <Dimmer
      active={false}
    >
      <DimmerInner
        active={false}
      >
        <div
          className="ui dimmer"
          onClick={[Function]}
        >
          <div
            className="content"
          >
            <Loader>
              <div
                className="ui loader"
              />
            </Loader>
          </div>
        </div>
      </DimmerInner>
    </Dimmer>
    <div
      className="plotly-chart"
      style={
        Object {
          "height": "100%",
          "marginBottom": 5,
          "width": "100%",
        }
      }
    />
  </div>
</PlotlyChart>
`;

exports[`PlotlyChart Should match existing snapshot when given empty data. 1`] = `
<PlotlyChart
  config={Object {}}
  data={Array []}
  height="100%"
  layout={Object {}}
  showLoader={true}
  width="100%"
>
  <div>
    <Dimmer
      active={true}
    >
      <DimmerInner
        active={true}
      >
        <div
          className="ui active transition visible dimmer"
          onClick={[Function]}
        >
          <div
            className="content"
          >
            <Loader>
              <div
                className="ui loader"
              />
            </Loader>
          </div>
        </div>
      </DimmerInner>
    </Dimmer>
    <div
      className="plotly-chart"
      style={
        Object {
          "height": "100%",
          "marginBottom": 5,
          "width": "100%",
        }
      }
    />
  </div>
</PlotlyChart>
`;

exports[`PlotlyChart Should match existing snapshot when given sample data. 1`] = `
<PlotlyChart
  config={Object {}}
  data={
    Array [
      Object {
        "marker": Object {
          "color": "blue",
        },
        "mode": "markers",
        "type": "pointcloud",
        "xy": Float32Array [
          1,
          2,
          3,
          4,
        ],
      },
    ]
  }
  height="100%"
  layout={Object {}}
  showLoader={true}
  width="100%"
>
  <div>
    <Dimmer
      active={false}
    >
      <DimmerInner
        active={false}
      >
        <div
          className="ui dimmer"
          onClick={[Function]}
        >
          <div
            className="content"
          >
            <Loader>
              <div
                className="ui loader"
              />
            </Loader>
          </div>
        </div>
      </DimmerInner>
    </Dimmer>
    <div
      className="plotly-chart"
      style={
        Object {
          "height": "100%",
          "marginBottom": 5,
          "width": "100%",
        }
      }
    />
  </div>
</PlotlyChart>
`;
